sockeye/training.py

```diff
@@ -393,15 +402,31 @@ def _fit(self,
                     checkpoint_state = CheckpointState(checkpoint=train_state.checkpoint, metric_val=m_val)
                     optimizer.pre_update_checkpoint(checkpoint_state)
 
+                # learning rate adjustment
                 if self.lr_scheduler is not None:
-                    self.lr_scheduler.new_evaluation_result(has_improved)
+                    lr_adjusted = self.lr_scheduler.new_evaluation_result(has_improved)
+                    if lr_adjusted and not has_improved:
+                        if lr_decay_param_reset:
+                            logger.info("Loading parameters from last best checkpoint: %d", best_checkpoint)
+                            self._load_params(output_folder, best_checkpoint)
+                        if lr_decay_opt_states_reset == C.LR_DECAY_OPT_STATES_RESET_INITIAL:
+                            logger.info("Resetting optimizer states to initial")
+                            self.load_optimizer_states(os.path.join(output_folder, C.OPT_STATES_INITIAL))
+                        elif lr_decay_opt_states_reset == C.LR_DECAY_OPT_STATES_RESET_BEST:
+                            logger.info("Resetting optimizer states to last best checkpoint: %d", best_checkpoint)
+                            self.load_optimizer_states(os.path.join(output_folder, C.OPT_STATES_BEST))
 
                 if has_improved:
-                    best_path = os.path.join(output_folder, C.PARAMS_BEST_NAME)
-                    if os.path.lexists(best_path):
-                        os.remove(best_path)
-                    actual_best_fname = C.PARAMS_NAME % best_checkpoint
-                    os.symlink(actual_best_fname, best_path)
+                    best_params_path = os.path.join(output_folder, C.PARAMS_BEST_NAME)
+                    if os.path.lexists(best_params_path):
+                        os.remove(best_params_path)
+                    actual_best_params_fname = C.PARAMS_NAME % best_checkpoint
+                    os.symlink(actual_best_params_fname, best_params_path)
+                    if lr_decay_opt_states_reset == C.LR_DECAY_OPT_STATES_RESET_BEST:
+                        best_opt_states_fname = os.path.join(output_folder, C.OPT_STATES_BEST)
+                        if os.path.exists(best_opt_states_fname):
+                            os.remove(best_opt_states_fname)
+                        self.save_optimizer_states(best_opt_states_fname)
```

Comment:
Minor: We will leave the optimizer state file in the output directory.

Reply:
good point, I will add deletion at the end and the merge.